File: src/main/java/net/dumbdogdiner/dogcore/util/LinkedQueue.java
package net.dumbdogdiner.dogcore.util;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

/**
 * A single-ended queue backed by a doubly-linked list of nodes.
 * All operations are O(1).
 * @param <E> The type of data to store in the queue.
 */
public final class LinkedQueue<E> {
    /**
     * An implementation of a node.
     * @param <E> The type of data stored in this node.
     */
    public static final class Node<E> {
        /** The value contained in the node. */
        private final E value;
        /** The previous node. */
        private @Nullable Node<E> prev;
        /** The next node. */
        private @Nullable Node<E> next = null;

        /** Prevents accidentally removing a node twice. */
        private boolean inQueue = true;

        private Node(final E nodeValue, @Nullable final Node<E> prevNode) {
            value = nodeValue;
            prev = prevNode;
        }

        @Override
        public int hashCode() {
            if (value != null) {
                return value.hashCode();
            }
            return 0;
        }

        /**
         * Nodes are only equal by identity.
         * @param other The other object to test equality with.
         * @return True if equal.
         */
        @Override
        public boolean equals(final Object other) {
            return this == other;
        }

        /**
         * Get the value stored in this node.
         * @return The value in this node.
         */
        public E getValue() {
            return value;
        }
    }

    /** The head of the list. */
    private @Nullable Node<E> head = null;

    /** The tail of the list. */
    private @Nullable Node<E> tail = null;

    /**
     * Push a new node to the back of the queue.
     * @param value The value to push.
     * @return The node at the back of the queue which contains the value.
     */
    public synchronized Node<E> push(final E value) {
        var newNode = new Node<>(value, tail);

        if (head == null) {
            head = newNode;
        }

        if (tail != null) {
            tail.next = newNode;
        }
        tail = newNode;
        return newNode;
    }

    /**
     * Get the node at the front of the queue.
     * @return The node at the front of the queue, or null.
     */
    public synchronized @Nullable Node<E> peek() {
        return head;
    }

    /**
     * Remove a node from the queue. Element order is maintained.
     * @param node The node to remove.
     */
    public synchronized void remove(@NotNull final Node<E> node) {
        if (!node.inQueue) {
            // node isn't a part of the queue, so don't do anything
            return;
        }

        var prev = node.prev;
        var next = node.next;

        if (prev != null) {
            prev.next = next;
        } else {
            head = next;
        }

        if (next != null) {
            next.prev = prev;
        } else {
            tail = prev;
        }

        node.prev = null;
        node.next = null;
        node.inQueue = false;
    }
}

File: src/main/kotlin/net/dumbdogdiner/dogcore/util/LinkedQueue.kt
package net.dumbdogdiner.dogcore.util

/**
 * A single-ended queue backed by a doubly-linked list of nodes. All operations are O(1).
 */
class LinkedQueue<E> {
    /** The public details of nodes. */
    sealed interface Node<E> {
        /** The value contained in the node. */
        val value: E
    }

    /** An implementation of a node. */
    private class NodeImpl<E>(
        /** The value contained in the node. */
        override val value: E,
        /** The previous node. */
        var prev: NodeImpl<E>?,
        /** The next node. */
        var next: NodeImpl<E>?,
        /** Prevents accidentally removing a node twice. */
        var inQueue: Boolean = true
    ) : Node<E> {
        override fun hashCode() = value.hashCode()

        /** Nodes are only equal by identity. */
        override fun equals(other: Any?) = this === other
    }

    /** The head of the list. */
    private var head: NodeImpl<E>? = null

    /** The tail of the list. */
    private var tail: NodeImpl<E>? = null

    /** Push a new node to the back of the queue. */
    fun push(value: E): Node<E> {
        val newNode = NodeImpl(value, tail, null)

        if (head == null) {
            head = newNode
        }

        tail?.let {
            it.next = newNode
        }

        tail = newNode
        return newNode
    }

    /** Get the node at the front of the queue. */
    fun peek(): Node<E>? {
        return head
    }

    /** Remove a node from the queue. Element order is maintained. */
    fun remove(node: Node<E>) {
        val nodeImpl = node as NodeImpl<E>
        if (!nodeImpl.inQueue) {
            // node isn't a part of the queue, so don't do anything
            return
        }

        nodeImpl.prev?.let {
            it.next = nodeImpl.next
        } ?: run {
            head = nodeImpl.next
        }

        nodeImpl.next?.let {
            it.prev = nodeImpl.prev
        } ?: run {
            tail = nodeImpl.prev
        }

        nodeImpl.prev = null
        nodeImpl.next = null
        nodeImpl.inQueue = false
    }
}
